File: src/main/java/combat_object/combatant/LifeStatus.java
package combat_object.combatant;

import __main.Main;
import lombok.*;

public class LifeStatus {

    @Getter private Status status = Status.ALIVE;
    private int deathSuccesses = 0, deathFails = 0;

    /**
     * Takes the param death save roll. Checks for values > 20 or < 1 and
     * increments either death successes or fails according to the roll.
     */
    public void rollDeathSave(int d20Roll) {
        if (d20Roll > 20 || d20Roll < 1) throw new IndexOutOfBoundsException();

        if (d20Roll >= 10)
            deathSuccesses++;
        else
            deathFails++;

        if (deathSuccesses == 3) {
            status = Status.ALIVE;
            deathSuccesses = 0;
            deathFails = 0;
        } else if (deathFails == 3) {
            status = Status.DEAD;
        }

        Main.refreshUI();
    }

    public boolean isConscious() {
        return status == Status.ALIVE;
    }

    public void setUnconscious() {
        status = Status.UNCONSCIOUS;
    }

    public boolean isAlive() {
        return status != Status.DEAD;
    }

    @Override
    public String toString() {
        return String.format("%d - %d Saves", deathSuccesses, deathFails);
    }

    public enum Status {ALIVE, UNCONSCIOUS, DEAD}
}

File: src/main/dnd_combat/character_info/LifeStatus.kt
package character_info

import __main.Main

class LifeStatus {
    enum class Status {
        ALIVE, UNCONSCIOUS, DEAD
    }

    private var thisStatus: Status = Status.ALIVE

    private var deathSuccesses = 0
    private var deathFails = 0

    /**
     * takes the param death save roll. Checks for values > 20 or < 1 and
     * increments either death successes or fails according to the roll.
     * Calls updateLifeStatus() to handle death/resurrection
     */
    fun rollDeathSave(d20Roll: Int) {
        if (d20Roll > 20 || d20Roll < 1) {
            throw IndexOutOfBoundsException()
        }
        if (d20Roll > 10) {
            deathSuccesses++
        } else {
            deathFails++
        }

        if (deathSuccesses == 3) {
            thisStatus = Status.ALIVE
        } else if (deathFails == 3) {
            thisStatus = Status.DEAD
        }

        Main.logAction()
    }

    fun status(): Status {
        return thisStatus
    }

    val isConscious: Boolean
        get() = thisStatus == Status.ALIVE

    fun setUnconscious() {
        thisStatus = Status.UNCONSCIOUS
    }

    val isAlive: Boolean
        get() = thisStatus != Status.DEAD

    override fun toString(): String {
        return String.format("Defeated (%d-%d)", deathSuccesses, deathFails)
    }
}